### PR TITLE
 [NEW] Image preview as 32x32 base64 jpeg

### PR DESCRIPTION
**Pull request by jorgeluisrezende**
@RocketChat/core 

Open #5563

Added a image preview as small base64 50x50 pictures embedded on the msg object.

I just made a function, and had call it on the sendFileMessage method, and i fixed the build errors.
I open another PR because my old branch was in conflict, at this time i did some build tests.

### CODE REVIEW

**Comment by jorgeluisrezende** [comment]
@rodrigok, now it is all ok!

File: packages/rocketchat-file-upload/server/methods/sendFileMessage.js [hunk 1]
```diff
@@ -32,42 +32,46 @@ Meteor.methods({
...
+			}, msgData);
 
-		msg = Meteor.call('sendMessage', msg);
+			msg = Meteor.call('sendMessage', msg);
 
-		Meteor.defer(() => RocketChat.callbacks.run('afterFileUpload', { user, room, message: msg }));
+			Meteor.defer(() => RocketChat.callbacks.run('afterFileUpload', { user, room, message: msg }));
 
-		return msg;
+			return msg;
```
Comment (rodrigok) [inline packages/rocketchat-file-upload/server/methods/sendFileMessage.js]: This is wrong, now the method will have nothing returned

File: packages/rocketchat-file-upload/server/methods/sendFileMessage.js [hunk 2]
```diff
@@ -32,42 +32,46 @@ Meteor.methods({
 			title_link_download: true
 		};
 
-		if (/^image\/.+/.test(file.type)) {
-			attachment.image_url = fileUrl;
-			attachment.image_type = file.type;
-			attachment.image_size = file.size;
-			if (file.identify && file.identify.size) {
-				attachment.image_dimensions = file.identify.size;
+		Meteor.wrapAsync(FileUpload.resizeImagePreview(file, Meteor.bindEnvironment(function(base64Preview) {
```
Comment (rodrigok) [inline packages/rocketchat-file-upload/server/methods/sendFileMessage.js]: You could check if file is image and call this only for images, makes more sense to me.

File: packages/rocketchat-file-upload/server/lib/FileUpload.js [hunk 3]
```diff
@@ -77,7 +78,20 @@ Object.assign(FileUpload, {
...
+		const transformer = sharp().resize(50, 50).max().toBuffer(function(err, out) {
+			if (err) { throw err; }
+			callback(out.toString('base64'));
+		});
+		if (/^image\/.+/.test(file.type)) {
+			image.pipe(transformer);
+		} else {
+			callback();
+		}
+	},
```
Comment (rodrigok) [inline packages/rocketchat-file-upload/server/lib/FileUpload.js]: You can create a new method like `resizeImagePreviewSync` that does not receive the callback and do the wrapAsync inside

Comment (jorgeluisrezende) [reply packages/rocketchat-file-upload/server/lib/FileUpload.js]: ok, i will do that!

Comment (jorgeluisrezende) [reply packages/rocketchat-file-upload/server/lib/FileUpload.js]: I tried to do that, but the file does not finish to send, might i'm doing something wrong.
I think the Bradley suggestion that uses Promises is nice, but i don't know if you like that..
What do you think?

File: packages/rocketchat-ui-message/client/renderMessageBody.js [hunk 4]
```diff
@@ -9,7 +9,6 @@ renderMessageBody = function(msg) {
 	}
 
 	const message = RocketChat.callbacks.run('renderMessage', msg);
-
```
Comment (rodrigok) [inline packages/rocketchat-ui-message/client/renderMessageBody.js]: Do not introduce unnecessary changes.

File: packages/rocketchat-file-upload/server/lib/FileUpload.js [hunk 5]
```diff
@@ -89,7 +90,18 @@ Object.assign(FileUpload, {
 		}));
 		return future.wait();
 	},
+	resizeImagePreview(file) {
+		return new Promise(function(resolve, reject) {
+			const image = FileUpload.getStore('Uploads')._store.getReadStream(file._id, file);
 
+			const transformer = sharp().resize(50, 50).max().toBuffer(function(err, out) {
+				if (err) { reject(err); }
```
Comment (rodrigok) [inline packages/rocketchat-file-upload/server/lib/FileUpload.js]: Without a return you will reject and resolve the promise right?

File: packages/rocketchat-file-upload/server/methods/sendFileMessage.js [hunk 6]
```diff
@@ -39,6 +39,7 @@ Meteor.methods({
 			if (file.identify && file.identify.size) {
 				attachment.image_dimensions = file.identify.size;
 			}
+
```
Comment (rodrigok) [inline packages/rocketchat-file-upload/server/methods/sendFileMessage.js]: You could execute the resize here using `await` to keep the flow sync

File: packages/rocketchat-file-upload/server/methods/sendFileMessage.js [hunk 7]
```diff
@@ -64,10 +65,18 @@ Meteor.methods({
...
+		isImage(file)
+			? FileUpload.resizeImagePreview(file)
+				.then(r => msg.attachments[0].imagePreview = r)
+				.then(sendMessage)
+			: sendMessage();
 
 		Meteor.defer(() => RocketChat.callbacks.run('afterFileUpload', { user, room, message: msg }));
 
 		return msg;
+
```
Comment (rodrigok) [inline packages/rocketchat-file-upload/server/methods/sendFileMessage.js]: Do not introduce unnecessary changes

File: packages/rocketchat-ui-message/client/renderMessageBody.js (hunk 4 above)
Comment (jorgeluisrezende) [reply packages/rocketchat-ui-message/client/renderMessageBody.js]: sorry, it was a mistake when i was making tests..

**Comment by jorgeluisrezende** [comment]
@rodrigok, now i guess it is ok.

File: packages/rocketchat-file-upload/server/methods/sendFileMessage.js [hunk 8]
```diff
@@ -24,50 +24,53 @@ Meteor.methods({
...
 		const fileUrl = `/file-upload/${ file._id }/${ encodeURI(file.name) }`;
 
-		const attachment = {
-			title: file.name,
-			type: 'file',
-			description: file.description,
-			title_link: fileUrl,
-			title_link_download: true
-		};
+		(async function() {
```
Comment (rodrigok) [inline packages/rocketchat-file-upload/server/methods/sendFileMessage.js]: Did you try to remove this `wrapper`? AFAIK all meteor methods are async by default allowing you to call await inside.

Comment (jorgeluisrezende) [reply packages/rocketchat-file-upload/server/methods/sendFileMessage.js]: I had tested a lot of ways to do it, when i tryed to call await without the wrapper it was failing in the build...

**Comment by rodrigok** [comment]
@sampaiodiego It is mainly for mobile guys right now, once merged I'll create a new issue for UI guys to start using it.
